csrc/attention/attention_kernels.cu

```diff
@@ -178,7 +178,7 @@ __global__ void single_query_cached_kv_attention_kernel(
       // This includes a reduction across the threads in the same thread group.
       float qk = scale * Qk_dot<scalar_t, THREAD_GROUP_SIZE>::dot(q_vecs[thread_group_offset], k_vecs);
       // Add the ALiBi bias if slopes are given.
-      qk += (alibi_slope != 0) ? alibi_slope * (token_idx - context_len) : 0;
+      qk += (alibi_slope != 0) ? alibi_slope * (token_idx - context_len + 1) : 0;
```

Comment:
Does this affect the attention output?
I was actually considering changing it to
```suggestion
      qk += (alibi_slope != 0) ? alibi_slope * token_idx;
```
Because they are mathematically the same.

Reply:
According to @Oliver-ss, this indeed changes the model outputs. The version I posted is what he suggested that matches Pytorch results. Later @Oliver-ss will update the test case in this PR.

Reply:
In my experience, the current implementation also matches the HF transformer's outputs when using FP32 (I used bloom-560m) while it doesn't perfectly match HF when using FP16. So I believe it should be a numerical error.

Reply:
Hi, it is about the relative distance in ALibi. According to the paper, it should be:
![screenshot-20230906-145511](https://github.com/vllm-project/vllm/assets/44120206/ada5213c-df19-49a1-932f-9f864aca5478)

The alibi mask in multi query attention is correct. https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/attention.py#L355
```
bias = torch.arange(prompt_len)
bias = bias[None, :] - bias[:, None]
bias = bias.to(self.alibi_slopes.device)
```

In the latest alibi test case, the reference alibi mask is https://github.com/vllm-project/vllm/blob/main/tests/kernels/test_attention.py#L86
```
position_ids = torch.arange(context_len, device="cuda").int()
alibi_bias = (context_len - position_ids).float()
```
if the context_len is 5, the alibi bias is [5, 4, 3, 2, 1]

In the cuda kernel implementation
```
qk += (alibi_slope != 0) ? alibi_slope * (token_idx - context_len) : 0;
```
the alibi bias is [-5, -4, -3, -2, -1]

However, the correct alibi bias should be [-4, -3, -2, -1, 0]. 

As to the numerical error, make the alibi slope larger could make such error more obvious.

Reply:
@Oliver-ss, thanks for letting us know the bug in our test code when `USE_ALIBI` is True, After fixing the test code, the current kernel passed the tests.

> the cuda kernel alibi bias is [-5, -4, -3, -2, -1]. However, the correct alibi bias should be [-4, -3, -2, -1, 0].

To my understanding, they are equal because of the following softmax, right? I'm good with this change as it is more aligned with the attention bias we use with xformers. However, still there is a concern that different models may still use different definitions of ALiBi. Fore example, in [BLOOM](https://github.com/huggingface/transformers/blob/172f42c512e1bf32554ef910fe82f07916b4d4af/src/transformers/models/bloom/modeling_bloom.py#L125), it seems the model uses `[0, 1, 2, 3, 4]` instead of `[-4, -3, -2, -1, 0]`.

Reply:
1. Actually they are not the same. For the decoding phase:`attn_out = softmax(qK^T+m([-4, -3, -2, -1, 0]))`. It would result in different attention score with different relative distance. Sometimes, because the alibi bias is relatively small so that even if it is incorrect, the output token will not change. (Just think about the test case, if make the absolute tolerance 1e-2, it could pass even if the original version use opposite distance.)

2. Yes, for Bloom and Baichuan(need to be checked later), they seems use [0, 1, 2, 3, 4] instead of [-4, -3, -2, -1, 0]. So the implementation in vLLM of Bloom might also be incorrect.

Reply:
I see. Yes, it is mathematically the same. Keep consistent with the the attn bias in xfomers would help avoid confusion.